src/armscan_env/volumes/slicing.py

```diff
@@ -0,0 +1,211 @@
+        """Get the angles from a rotation matrix."""
+        # Extract the angles from the rotation matrix
+        th_x2 = np.arcsin(rotation_matrix[2, 1])
+        th_z1 = np.arcsin(rotation_matrix[1, 0])
+
+        # Convert the angles to degrees
+        th_z1 = np.rad2deg(th_z1)
+        th_x2 = np.rad2deg(th_x2)
+
+        return np.array([th_z1, th_x2])
+
+    def transform_action(self, relative_action: ManipulatorAction) -> ManipulatorAction:
+        """Transform an action to be relative to the new coordinate system."""
+        volume_transform_matrix = self.get_transform()
+
+        action_matrix = EulerTransform(relative_action).get_transform()
+        new_action_matrix = np.dot(
+            np.linalg.inv(volume_transform_matrix),
+            action_matrix,
+        )  # 1_A_s = 1_T_0 * 0_A_s
+
+        # new_action_translation = action_matrix[:2, 3] - volume_transform_matrix[:2, 3]
+        new_action_rotation = self.get_angles_from_rotation_matrix(new_action_matrix[:3, :3])
+        new_action_translation = new_action_matrix[:2, 3]
+
+        transformed_action = ManipulatorAction(
+            rotation=new_action_rotation,
+            translation=new_action_translation,
+        )
+
+        log.debug(
+            f"Random transformation: {self.action}\n"
+            f"Original action: {relative_action}\n"
+            f"Transformed action: {transformed_action}\n",
+        )
+
+        return transformed_action
+
+
+class TransformedVolume(sitk.Image):
+    def __init__(self, volume: sitk.Image, action: ManipulatorAction):
+        super().__init__(volume)
+        self.transformation_action = action
+
+
+def transform_volume(
+    volume: sitk.Image,
+    action: ManipulatorAction,
+) -> TransformedVolume:
+    """Trasnform a 3D volume with arbitrary rotation and translation.
+
+    :param volume: 3D volume to be transformed
+    :param action: action to transform the volume
+    :return: the sliced volume.
+    """
+    origin = np.array(volume.GetOrigin())
+    euler_transform = EulerTransform(action, origin)
+    eul_tr = euler_transform.get_transform()
+
+    # Define plane's coordinate system
+    e1 = eul_tr[0][:3]
+    e2 = eul_tr[1][:3]
+    e3 = eul_tr[2][:3]
+    img_o = eul_tr[:, -1:].flatten()[:3]  # origin of the image plane
+
+    direction = np.stack([e1, e2, e3], axis=0).flatten()
+
+    resampler = sitk.ResampleImageFilter()
+    spacing = volume.GetSpacing()
+
+    resampler.SetOutputDirection(direction.tolist())
+    resampler.SetOutputOrigin(img_o.tolist())
+    resampler.SetOutputSpacing(spacing)
+    resampler.SetSize(volume.GetSize())
+    resampler.SetInterpolator(sitk.sitkNearestNeighbor)
+
+    # Resample the volume on the arbitrary plane
+    transformed_volume = resampler.Execute(volume)
+    # Todo: hack --> needed to break rotation dependency
+    return TransformedVolume(transformed_volume, action)
+
+
+def slice_volume(
+    volume: sitk.Image | TransformedVolume,
```

Comment:
You should only ever used TransformedVolume instead of bare `sitk.Image` in your code now. The `load_image` function should load an instance of it and there should be a default value for the transform (probably `None` will do)

Reply:
Obsolete after my last commits